File: Algo/Leetcode/494TargetSum.cpp
// dfs. O(2^n) time
class Solution {
public:
    int count = 0;
    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        if(n==0) return 0;
        dfs(nums, 0, 0, target);
        return count;
    }

    void dfs(vector<int>& nums, int x, int sum, int target){
        if(x == nums.size()) {
            if(sum == target) count++;
            return;
        }

        dfs(nums, x+1, sum+nums[x], target);
        dfs(nums, x+1, sum-nums[x], target);
    }
};

// dp.
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int sum = 0;
        for(auto c:nums) sum+=c;
        sum = sum+target;
        if(sum<0 || sum%2) return 0;
        sum /= 2;

        vector<int> dp(sum+1);
        dp[0] = 1;
        for(auto i:nums){
            for(int j=sum; j>=i; j--)
                dp[j] += dp[j-i];
        }
        return dp[sum];
    }
};

File: Algo/Leetcode/494TargetSum.py
class Solution(object):

    def findTargetSumWays(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        s = sum(nums) + target
        if s < 0 or s % 2 != 0:     # sum must be pos
            return 0
        s = s / 2

        dp = [[0 for _ in range(s+1)] for _ in range(n+1)] # (n+1, s+1)
        dp[0][0] = 1    # not choose anyone, and sum 0

        for i in range(1, n+1):
            for j in range(0, s+1):
                dp[i][j] = (dp[i-1][j-nums[i-1]] if nums[i-1] <= j else 0) + dp[i-1][j]

        return dp[-1][-1]